chore(compiler): emit ([] as any[]) instead of purely []

rkirov opened the pull request:
**What kind of change does this PR introduce?** (check one with "x")

```
[] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

rkirov commented:
blocked on independently broken build

vicb commented on modules/@angular/compiler-cli/integrationtest/src/bootstrap.ts:
```diff
@@ -10,4 +10,4 @@ import {platformBrowser} from '@angular/platform-browser';
 import {MainModuleNgFactory} from './module.ngfactory';
 
-MainModuleNgFactory.create(null).instance.appRef.bootstrap(BasicComp);
+MainModuleNgFactory.create(null as any).instance.appRef.bootstrap(BasicComp);
```
why do we need this `null as any`  ? (wondering as I have also seen this in codegen)

/cc @tbosch @pkozlowski-opensource

alexeagle replied:
Because NgFactory.create doesn't accept the `null` type, and it fails the type-check with `--strictNullChecks` which we use globally in g3

kenhahn85 commented on modules/@angular/forms/src/model.ts:
```diff
@@ -473,12 +478,7 @@ export class FormControl extends AbstractControl {
-    emitModelToViewChange?: boolean,
-    emitViewToModelChange?: boolean
-  } = {}): void {
+  setValue(value: any, {onlySelf, emitEvent, emitModelToViewChange, emitViewToModelChange}: any = {}): void {
```
You can preserve type safety with strict null checking by using something like:

```
options: {
    onlySelf?: boolean,		
    emitEvent?: boolean,		
    emitModelToViewChange?: boolean,		
    emitViewToModelChange?: boolean		
 } = {};
```

opensrcken replied:
You can preserve type safety with strict null checking by using something like:

```
options: {
    onlySelf?: boolean,		
    emitEvent?: boolean,		
    emitModelToViewChange?: boolean,		
    emitViewToModelChange?: boolean		
 } = {};
```

rkirov replied:
This was working around a bug https://github.com/Microsoft/TypeScript/issues/10078 and intentionally reducing type safety. No longer needed thankfully.

opensrcken commented:
Would be great to get this in. 👍

rkirov committed:
chore(compiler): emit ([] as any[]) instead of purely [].

In SNC mode `[]` has type of never[], so we cast it to any[] to
typecheck correctly see
https://github.com/Microsoft/TypeScript/issues/10479.

This is temporary workaround, until we fully
migrate the framework to TS 2.0 and strictNullChecks.

rkirov commented:
I appears that the framework has been greatly decoupled from ngc emitted factories, so this PR has similarly decreased in size.

Note that if one attempts to turn on SNC for the framework itself a lot of the same changes will need to be made (feel free to ping me when attempting it).

rkirov commented:
Further replaced by https://github.com/angular/angular/pull/11846 due to CircleCI caching issues :(
